docs/prompting/decomposition/skeleton_of_thought.md

```diff
@@ -1,7 +1,126 @@
+# from multiprocessing import Pool
+
+client = instructor.from_openai(OpenAI())
+
+
+class Point(BaseModel):
+    index: int
+    description: str
+
+
+class Skeleton(BaseModel):
+    points: list[Point]
+
+
+class Response(BaseModel):
+    response: str
+
+
+def get_skeleton(question):
+    return client.chat.completions.create(
+        model="gpt-4o",
+        response_model=Skeleton,
+        messages=[
+            {
+                "role": "user",
+                "content": f"""
+                You’re an organizer responsible for only giving the skeleton (not the full content) for answering the question.
+                Provide the skeleton in a list of points (numbered 1., 2., 3., etc.) to answer the question.
+                Instead of writing a full sentence, each skeleton point should be very short with only 3∼5 words.
+                Generally, the skeleton should have 3∼10 points.
+
+                Now, please provide the skeleton for the following question.
+
+                <question>
+                {question}
+                </question>
+
+                Skeleton:
+                """,
+            }
+        ],
+    )
+
+
+def expand_point(input):
+    question, skeleton, point_index = input
+
+    return client.chat.completions.create(
+        model="gpt-4o",
+        response_model=Response,
+        messages=[
+            {
+                "role": "user",
+                "content": f"""
+                You’re responsible for continuing the writing of one and only one point in the overall answer to the following question.
+
+                <question>
+                {question}
+                </question>
+
+                The skeleton of the answer is:
+
+                <skeleton>
+                {skeleton}
+                </skeleton>
+
+                Continue and only continue the writing of point {point_index}.
+                Write it **very shortly** in 1∼2 sentence and do not continue with other points!
+                """,
+            }
+        ],
+    ).response
+
+
+if __name__ == "__main__":
+    query = "Compose an engaging travel blog post about a recent trip to Hawaii, highlighting cultural experiences and must-see attractions."
+
+    # Step 1: Get the skeleton
+    skeleton = get_skeleton(query)
+
+    for point in skeleton.points:
+        print(point)
+        #> index=1 description='Introduction to Hawaii trip'
+        #> index=2 description='First impressions'
+        #> index=3 description='Cultural insights'
+        #> index=4 description='Traditional food tasting'
+        #> index=5 description='Visit to historical sites'
+        #> index=6 description='Beach experiences'
+        #> index=7 description='Adventure activities'
+        #> index=8 description='Local festivals and events'
+        #> index=9 description='Meeting locals'
+        #> index=10 description='Conclusion and reflections'
+
+    # Step 2: Expand on each point in parallel
+    # with Pool() as pool:
```

Comment:
suggestions on handling pytest with multiprocessing?

Reply:
lets not use multiprocessing but asyncio.gather https://python.useinstructor.com/blog/2023/11/13/learn-async/?h=asyncio